src/c/core/session/session.c

```diff
@@ -520,22 +520,32 @@ void read_submessage_fragment(uxrSession* session, ucdrBuffer* submessage, uxrSt
     //TODO
 }
 
-void read_submessage_heartbeat(uxrSession* session, ucdrBuffer* submessage, uxrStreamId stream_id)
+void read_submessage_heartbeat(uxrSession* session, ucdrBuffer* submessage)
 {
-    uxrInputReliableStream* stream = uxr_get_input_reliable_stream(&session->streams, stream_id.index);
+    HEARTBEAT_Payload heartbeat;
+    uxr_deserialize_HEARTBEAT_Payload(submessage, &heartbeat);
+    uxrStreamId id = uxr_stream_id_from_raw(heartbeat.stream_id, UXR_INPUT_STREAM);
+
+    uxrInputReliableStream* stream = uxr_get_input_reliable_stream(&session->streams, id.index);
     if(stream)
     {
-        uxr_read_heartbeat(stream, submessage);
-        write_submessage_acknack(session, stream_id);
+        uxr_process_heartbeat(stream, heartbeat.first_unacked_seq_nr, heartbeat.last_unacked_seq_nr);
+        write_submessage_acknack(session, id);
     }
 }
 
-void read_submessage_acknack(uxrSession* session, ucdrBuffer* submessage, uxrStreamId stream_id)
+void read_submessage_acknack(uxrSession* session, ucdrBuffer* submessage)
 {
-    uxrOutputReliableStream* stream = uxr_get_output_reliable_stream(&session->streams, stream_id.index);
+    ACKNACK_Payload acknack;
+    uxr_deserialize_ACKNACK_Payload(submessage, &acknack);
+    uxrStreamId id = uxr_stream_id_from_raw(acknack.stream_id, UXR_INPUT_STREAM);
+
+    uxrOutputReliableStream* stream = uxr_get_output_reliable_stream(&session->streams, id.index);
     if(stream)
     {
-        uxr_read_acknack(stream, submessage);
+        uint16_t nack_bitmap = (uint16_t)(((uint16_t)acknack.nack_bitmap[0] << 8) +
```

Comment:
Ok, first of all, I do know how to fix it properly. 
Previously this sentence was in OutputReliableBuffer, and the otherwise logic was in InputReliableBuffer. Now, one is in Session and the otherwise in InputReliableBuffer. Maybe it can be confused and should be at the same logic level. I do not know if the solution is to upload the InputReliableBuffer logic here, to session, or try to lower this logic to OutputReliableStream. Maybe to put into the OutputReliableBuffer is better because is a low-level logic that the session should not know. By other side, maybe it can be a good moment to clean the input and output reliable entities of the serialization code dependency, and get these entities agnostic to that package. Anyway, I think that is important that the opposite logics are at the same level.

Reply:
I moved the logic from Stream to Session in order to be compliant with other submessages as `STATUS`. It looks like more robus from my point of view, in fact I removed the `xrce_protocol` dependencies from Streams. What is your opinion?

Reply:
Perfect!! less dependencies. Could you adapt the `core_class.uml` file of internal docs to this changes? (the lines can be hidden to maintan the layout if it breaks, instead of remove it). 

As a thought, I think that session is having too many concepts, and it will need a refactor in the future into different entities to improve the organization.

Reply:
Updated internal documentation.